### ISSUE TITLE
ListingTable cannot handle partition evolution

### ISSUE THREAD

**Issue by adriangb** [apache/datafusion#13270]
### Describe the bug

With CSV:

```shell
echo "a,b\n1,2" > data1.csv
mkdir a=2
echo "b\n3" > a=2/data2.csv
datafusion-cli
> SELECT * FROM '**/*.csv';
Arrow error: Csv error: incorrect number of fields for line 1, expected 2 got 1
```

With Parquet:

```python
import os
import polars as pl

pl.DataFrame({'a': [1], 'b': [2]}).write_parquet('data1.parquet')
os.mkdir('a=2')
pl.DataFrame({'b': [3]}).write_parquet('a=2/data2.parquet')
```

```shell
datafusion-cli
> SELECT * FROM '**/*.parquet';
+---+---+
| b | a |
+---+---+
| 2 | 1 |
| 3 |   |
+---+---+
2 row(s) fetched.
Elapsed 0.055 seconds.
```

### Expected behavior

Partition evolution is handled and both cases return

```
+---+---+
| b | a |
+---+---+
| 2 | 1 |
| 3 | 2 |
+---+---+
```

### Additional context

Having played around quite a bit with ParquetExec and the SchemaAdapter machinery I think what should happen is:
- Partition values are on a per-file basis, in particular on each `PartitionedFile` and not on the `FileScanConfig`
- Partition values are passed into the SchemaAdapter machinery and for each file it decides if it needs to add a column generated from partition values or not

**Comment by adriangb** [comment]
cc @alamb I had promised you this a long time ago but only got around to it now

**Comment by alamb** [comment]
Thanks @adriangb

**Comment by zhuqi-lucas** [comment]
take

**Comment by logan-keede** [comment]
@adriangb my focus has been on refactoring `FileScanConfig` to move it out of core. I cant say I understand the internals that much, but I will look into it and mention it here if I find something relevant.

**Comment by zhuqi-lucas** [comment]
@adriangb Sorry for the delay, i am starting to investigate this issue this week.

**Comment by zhuqi-lucas** [comment]
First round investigation:

We need runtime to infer partitions result to overwrite the empty FileScanConfig table_partition_cols, i can't find a good way until now.


@adriangb Do you have any suggestions, how can we do this in current architecture?

**Comment by adriangb** [comment]
I think the fundamental issue is that the partition columns are specified on a per-exec basis via `FileScanConfig`. The only solutions I can think of are:
- Change the APIs to allow multiple `FileScanConfig`'s to be supplied. This bring about issues of making sure the output schemas all match so they can be unioned, etc.
- Move partition column generation into `SchemaAdapter`. The issue with this is that `SchemaAdapter` exists at a lower level than the concept of partition columns and it might be inappropriate to put that logic in there directly. But at the same time `FileScanConfig` and `ParquetExec` (recently folded into `ParquetDataSource`?) exist above the level of a single file, and partitioning can be as granular as a single file. I think the solution here would be to add hooks into `SchemaAdapter` to be able to handle missing columns so that the exec can inject information on how to generate partition columns from file paths. It could do that very dynamically on a per-file basis with no config or we could say that you have to pass the union of all columns that might be partition columns along with their field types and then if a file has only a subset of those that's okay, but we error or fill in nulls if we encounter a missing column that was not declared as a partition column.

**Comment by TheBuilderJR** [comment]
+1 I'm also blocked on this. It'd be nice if schema evolution could be a first class citizen in datafusion. It's been pretty painful/stressful running into schema evolution bugs with https://telemetry.sh. It feels like a ticking time bomb before a schema gets corrupted :(

**Comment by zhuqi-lucas** [comment]
Just noticed we have a solution for partition evolution, see details PR, may be we need some improvement based on it?

https://github.com/apache/datafusion/pull/12683/files

**Comment by TheBuilderJR** [comment]
@zhuqi-lucas here's a concrete repro for what I expect datafusion to be able to do but can't

```
use std::fs;
use std::sync::Arc;
use datafusion::prelude::*;
use datafusion::arrow::datatypes::{DataType, Field, Schema, TimeUnit};
use datafusion::arrow::record_batch::RecordBatch;
use datafusion::arrow::array::{Array, StringArray, StructArray, TimestampMillisecondArray};
use datafusion::datasource::listing::{ListingOptions, ListingTable, ListingTableConfig, ListingTableUrl};
use datafusion::datasource::file_format::parquet::ParquetFormat;
use datafusion::dataframe::DataFrameWriteOptions;

#[tokio::test]
async fn test_schema_evolution_with_compaction() -> Result<(), Box<dyn std::error::Error>> {
    let ctx = SessionContext::new();

    let schema1 = Arc::new(Schema::new(vec![
        Field::new("event", DataType::Utf8, false),
        Field::new("timestamp_utc", DataType::Timestamp(TimeUnit::Millisecond, None), false),
    ]));
    
    let batch1 = RecordBatch::try_new(
        schema1.clone(),
        vec![
            Arc::new(StringArray::from(vec!["event1"])),
            Arc::new(TimestampMillisecondArray::from(vec![1640995200000]))
        ]
    )?;

    let path1 = "test_data1.parquet";
    let _ = fs::remove_file(path1);
    
    let df1 = ctx.read_batch(batch1)?;
    df1.write_parquet(
        path1,
        DataFrameWriteOptions::default()
            .with_single_file_output(true)
            .with_sort_by(vec![col("timestamp_utc").sort(true, true)]),
        None
    ).await?;

    let schema2 = Arc::new(Schema::new(vec![
        Field::new("event", DataType::Utf8, false),
        Field::new("timestamp_utc", DataType::Timestamp(TimeUnit::Millisecond, None), false),
        Field::new("data", DataType::Struct(vec![
            Field::new("some_data", DataType::Utf8, false)
        ].into()), false),
    ]));

    let batch2 = RecordBatch::try_new(
        schema2.clone(),
        vec![
            Arc::new(StringArray::from(vec!["event2"])),
            Arc::new(TimestampMillisecondArray::from(vec![1641081600000])),
            Arc::new(StructArray::from(vec![(
                Arc::new(Field::new("some_data", DataType::Utf8, false)),
                Arc::new(StringArray::from(vec!["additional_data"])) as Arc<dyn Array>
            )]))
        ]
    )?;

    let path2 = "test_data2.parquet";
    let _ = fs::remove_file(path2);
    
    let df2 = ctx.read_batch(batch2)?;
    df2.write_parquet(
        path2,
        DataFrameWriteOptions::default()
            .with_single_file_output(true)
            .with_sort_by(vec![col("timestamp_utc").sort(true, true)]),
        None
    ).await?;

    let paths_str = vec![path1.to_string(), path2.to_string()];
    let config = ListingTableConfig::new_with_multi_paths(
        paths_str
            .into_iter()
            .map(|p| ListingTableUrl::parse(&p))
            .collect::<Result<Vec<_>, _>>()?
    )
        .with_schema(schema2.as_ref().clone().into())
        .infer(&ctx.state()).await?;

    let config = ListingTableConfig {
        options: Some(ListingOptions {
            file_sort_order: vec![vec![
                col("timestamp_utc").sort(true, true),
            ]],
            ..config.options.unwrap_or_else(|| ListingOptions::new(Arc::new(ParquetFormat::default())))
        }),
        ..config
    };

    let listing_table = ListingTable::try_new(config)?;
    ctx.register_table("events", Arc::new(listing_table))?;

    let df = ctx.sql("SELECT * FROM events ORDER BY event").await?;
    let results = df.clone().collect().await?;
    
    assert_eq!(results.len(), 2);

    let compacted_path = "test_data_compacted.parquet";
    let _ = fs::remove_file(compacted_path);

    df.write_parquet(
        compacted_path,
        DataFrameWriteOptions::default()
            .with_single_file_output(true)
            .with_sort_by(vec![col("timestamp_utc").sort(true, true)]),
        None
    ).await?;

    let new_ctx = SessionContext::new();
    let config = ListingTableConfig::new_with_multi_paths(vec![ListingTableUrl::parse(compacted_path)?])
        .with_schema(schema2.as_ref().clone().into())
        .infer(&new_ctx.state()).await?;
    
    let listing_table = ListingTable::try_new(config)?;
    new_ctx.register_table("compacted_events", Arc::new(listing_table))?;

    let df = new_ctx.sql("SELECT * FROM compacted_events ORDER BY event").await?;
    let compacted_results = df.collect().await?;
    
    assert_eq!(compacted_results.len(), 2);
    assert_eq!(results, compacted_results);

    let _ = fs::remove_file(path1);
    let _ = fs::remove_file(path2);
    let _ = fs::remove_file(compacted_path);

    Ok(())
}
```

**Comment by TheBuilderJR** [comment]
@zhuqi-lucas here's one current failure scenario with evolution: https://github.com/apache/datafusion/issues/14755

**Comment by adriangb** [comment]
Nowadays this returns `Error during planning: Found mixed partition values on disk [[], ["a"]]`
